File: backend/app/schemas/sales.py
from datetime import date, datetime

from pydantic import BaseModel


# --- Daily Sales Entry ---

class DailySalesInput(BaseModel):
    entry_date: date
    planned_sales: int


class DailySalesBatchInput(BaseModel):
    start_date: date
    end_date: date
    daily_sales: int


class DailySalesUpdate(BaseModel):
    planned_sales: int


# --- Inventory Override ---

class OverrideCreate(BaseModel):
    override_date: date
    override_value: int
    reason: str | None = None


class OverrideResponse(BaseModel):
    id: int
    override_date: date
    override_value: int
    reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Sales Plan ---

class SalesPlanCreate(BaseModel):
    plan_name: str
    sku: str | None = None
    asin: str | None = None
    start_date: date
    end_date: date
    initial_inventory: int = 0
    shipment_plan_id: int | None = None


class SalesPlanUpdate(BaseModel):
    plan_name: str | None = None
    sku: str | None = None
    asin: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    initial_inventory: int | None = None
    shipment_plan_id: int | None = None


class SalesPlanResponse(BaseModel):
    id: int
    plan_name: str
    sku: str | None
    asin: str | None
    start_date: date
    end_date: date
    initial_inventory: int
    shipment_plan_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SalesPlanListItem(BaseModel):
    id: int
    plan_name: str
    sku: str | None
    asin: str | None
    start_date: date
    end_date: date
    initial_inventory: int
    shipment_plan_id: int | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Calculation Results ---

class ArrivalDetail(BaseModel):
    unit_label: str
    quantity: int


class DailyCalculationResult(BaseModel):
    date: date
    opening_stock: int
    arrivals: int
    available_stock: int
    planned_sales: int
    actual_sales: int
    closing_stock: int
    is_stockout: bool
    has_override: bool
    arrival_details: list[ArrivalDetail] = []


class CalculationSummary(BaseModel):
    total_days: int
    total_planned_sales: int
    total_actual_sales: int
    stockout_days: int
    stockout_dates: list[date]
    ending_inventory: int


class CalculationResponse(BaseModel):
    sales_plan_id: int
    calculation_date: datetime
    summary: CalculationSummary
    daily_data: list[DailyCalculationResult]


# --- Turnover ---

class ShipmentTurnoverResult(BaseModel):
    unit_id: int
    unit_label: str
    region: str
    ship_date: date
    arrival_date: date
    quantity: int
    sold_quantity: int
    avg_turnover_days: float | None
    fully_sold: bool
    sell_through_date: date | None


class TurnoverResponse(BaseModel):
    sales_plan_id: int
    turnovers: list[ShipmentTurnoverResult]


# --- Chart Data ---

class ChartDataPoint(BaseModel):
    date: date
    opening_stock: int
    planned_sales: int
    actual_sales: int
    is_stockout: bool
    arrivals: int
    has_override: bool


class ChartDataResponse(BaseModel):
    sales_plan_id: int
    data: list[ChartDataPoint]


# --- Stockout Warning ---

class StockoutWarning(BaseModel):
    date: date
    planned_sales: int
    available_stock: int
    shortfall: int


class StockoutWarningResponse(BaseModel):
    sales_plan_id: int
    warnings: list[StockoutWarning]

File: frontend/src/types/sales.ts
export interface SalesPlanCreate {
  plan_name: string;
  sku?: string;
  asin?: string;
  start_date: string;
  end_date: string;
  initial_inventory: number;
  shipment_plan_id?: number | null;
}

export interface SalesPlanUpdate {
  plan_name?: string;
  sku?: string;
  asin?: string;
  start_date?: string;
  end_date?: string;
  initial_inventory?: number;
  shipment_plan_id?: number | null;
}

export interface SalesPlanResponse {
  id: number;
  plan_name: string;
  sku: string | null;
  asin: string | null;
  start_date: string;
  end_date: string;
  initial_inventory: number;
  shipment_plan_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface SalesPlanListItem {
  id: number;
  plan_name: string;
  sku: string | null;
  asin: string | null;
  start_date: string;
  end_date: string;
  initial_inventory: number;
  shipment_plan_id: number | null;
  created_at: string;
}

export interface DailySalesInput {
  entry_date: string;
  planned_sales: number;
}

export interface DailySalesBatchInput {
  start_date: string;
  end_date: string;
  daily_sales: number;
}

export interface OverrideCreate {
  override_date: string;
  override_value: number;
  reason?: string;
}

export interface OverrideResponse {
  id: number;
  override_date: string;
  override_value: number;
  reason: string | null;
  created_at: string;
}

export interface ArrivalDetail {
  unit_label: string;
  quantity: number;
}

export interface DailyCalculationResult {
  date: string;
  opening_stock: number;
  arrivals: number;
  available_stock: number;
  planned_sales: number;
  actual_sales: number;
  closing_stock: number;
  is_stockout: boolean;
  has_override: boolean;
  arrival_details: ArrivalDetail[];
}

export interface CalculationSummary {
  total_days: number;
  total_planned_sales: number;
  total_actual_sales: number;
  stockout_days: number;
  stockout_dates: string[];
  ending_inventory: number;
}

export interface CalculationResponse {
  sales_plan_id: number;
  calculation_date: string;
  summary: CalculationSummary;
  daily_data: DailyCalculationResult[];
}

export interface ShipmentTurnoverResult {
  unit_id: number;
  unit_label: string;
  region: string;
  ship_date: string;
  arrival_date: string;
  quantity: number;
  sold_quantity: number;
  avg_turnover_days: number | null;
  fully_sold: boolean;
  sell_through_date: string | null;
}

export interface TurnoverResponse {
  sales_plan_id: number;
  turnovers: ShipmentTurnoverResult[];
}

export interface StockoutWarning {
  date: string;
  planned_sales: number;
  available_stock: number;
  shortfall: number;
}
